### DISCUSSION TITLE
온보딩) 기능 분석하기

### DISCUSSION THREAD

**Discussion by Choi-JJunho** [BCSDLab/KOIN_API_V2#190] [에러 / 고민사항]
기존 코드에 구성되어있는 설정을 알아보고 아래 discussion으로 남겨주세요.

# `@Auth` 어노테이션은 어떻게 회원 아이디를 반환할까?

# `@IpAdress` 어노테이션은 어떻게 아이피 값을 받아올까?

# 테스트마다 데이터베이스 초기화가 이뤄지는데 어떤 과정을 거쳐 동작할까?

**Comment by duehee** [comment]
관련한 내용을 구글 문서를 통하여 정리해보았습니다.
https://docs.google.com/document/d/1IrhN9lV1Lsu9Hm7L3F0I9ZfSFc4HGiKTzEwbgPwPgq0/edit
어려웠던 부분도 있었어서, 작성하며 조금 이상한 설명이 있을 수도 있습니다..!

혹시나 제가 잘못 알고 있는 부분이 있다면 알려주시면 잘 학습해보겠습니다 :D

**Reply by Choi-JJunho** [reply]
정리를 잘하셨네요 👍
단순히 구글 문서로만 남겨두기엔 너무 아까운 정리내용인데 이참에 개인 블로그로 정리해보세요!

**Comment by dradnats1012** [comment]
[`@Auth`, `@IpAddress` 블로깅](https://velog.io/@dradnats1012/Auth-%EB%8F%99%EC%9E%91-%EB%B0%A9%EC%8B%9D)

[`DBInitializer` 블로깅](https://velog.io/@dradnats1012/DBInitializer)

주어진 키워드 관련해서 블로깅해봤습니다! 
틀린 부분 있으면 말씀해주세요

**Reply by Choi-JJunho** [reply]
Good 잘 정리해주셨네요 👍

**Comment by 20HyeonsuLee** [comment]
@Auth 실행 흐름
1. ExtractAuthenticationInterceptor클래스
    * preHandle메소드(HandlerInterceptor의 구현체로 요청이 들어오면 인터셉터로 사전처리한다.)
        1. 요청이 들어오면, ExtractAuthenticationInterceptor에 의해 토큰을 검증한다.
        2. 토큰이 유효하다면 AuthContext에 userId를 저장한다.
2. AuthArgumentResolver클래스
    * supportsParameter메소드(HandlerMethodArgumentResolver의 구현체로 파라미터로 어노테이션이 있는지 확인한다.)
        1. @Auto어노테이션이 있는지 확인한다.
    * resolveArgument메소드(HandlerMethodArgumentResolver의 구현체로 @Auth어노테이션이 있다면 실행된다.)
        1. preHandle메소드에서 설정한 AuthContext를 이용해서 사용자를 조회한다.
        2. 조회 결과에 따라 @Auth에서 지정한 권한인지 확인한 후 userId를 반환한다.

@IpAddress: 전체적인 흐름은 @Auth와 동일하다.

테스트마다 데이터 초기화
1. DBInitializer는 DB를 초기화해주는 클래스이다.
2. AcceptanceTest라는 추상클래스는 DBInitializer를 통해 DB를 초기화한다.
3. 테스트하는 클래스에서 AcceptanceTest를 상속받는다.
4. @BeforeEach처리된 함수가 테스트전에 실행된다.

알게된점: ExtractAuthenticationInterceptor클래스에서 정의한 AuthContext의 객체를 AuthArgumentResolver에서 사용하려면 AuthContext클래스에 @RequestScope를 붙이면 된다.
@RequestScope: 객체의 생명주기를 하나의 요청으로 설정한다. 때문에 하나의 요청으로 생성된 AuthContext객체를 AuthArgumentResolver, ExtractAuthenticationInterceptor에서 주입받아서 사용할 수 있다.

느낀점: 비기너 회고프로젝트를 했을 때 요청의 토큰에서 사용자 ID, IpAdress를 추출해야하는 경우가 있었다. 그때는 요청의 파라미터로 HttpServletRequest request를 받은 후 별도의 메소드를 작성하여 request로부터 토큰을 추출한 후 토큰으로부터 사용자의 정보를 추출했다. 그때의 방법과 지금의 어노테이션을 비교해보니 어노테이션으로 정의하는것이 확실히 가독성이 좋은것같다.

텍스트로 올리니 너무 번잡하네요..ㅎ 내일중으로 따로 정리해서 다시 올려보겠습니다. 그동안 피드백주세요!

**Reply by Choi-JJunho** [reply]
requestScope까지 알아봐주셨군요 👍
잘 정리해주셨네요~ 핵심적인 부분은 잘 이해하고계신것 같아 나머지 디테일한 부분은 굳이 이야기 안하겠습니다 (IPAdredss에서 ip 추출하는 방법과 같은..

**Comment by ImTotem** [comment]
[[Notion] KOIN_API_V2 기능 분석](https://imtotem.notion.site/KOIN_API_V2-5b27c3cb62bb432db683b26b1a1cdb45)

제가 잘못 알고있는 부분이 있다면 말씀해주세요!

**Reply by Choi-JJunho** [reply]
ArgumentResolver 선언과 등록까지 깔끔하게 정리해주셨네요~~
너무 잘 정리해주셨어요 👍👍

**Comment by seongjae6751** [comment]
https://velog.io/@seongjae6751/Custom-Annotation-만들기1-유저의-아이디-가져오기
https://velog.io/@seongjae6751/테스트를-위한-데이터-베이스-초기화

저도 블로그로 정리해봤습니다! 부족한 부분이 있다면 말씀해주시면 더 공부하고 알아보도록 하겠습니다.

**Reply by Choi-JJunho** [reply]
문제인식 - 해결방법 탐구의 순서대로 글을 정리해주셨네요~!
왜 이러한 구조가 필요한지 이해하고자하는 접근방식이 좋네요 👍
위에서 @20HyeonsuLee 가 정리한 내용 중 RequestScope에 대한 내용도 숙지해두면 좋을 것 같아요

DB 초기화도 깔끔하게 정리해주셨네요 👍

**Comment by Choon0414** [comment]
[@interface(커스텀 어노테이션)](https://intelligent-jade-f4a.notion.site/interface-995a0827b1ae4861a8d66e8a5229da03)

노션으로 정리했습니다. 아직 헷갈리고 모자란 부분이 많습니다. 많은 피드백 부탁드립니다! 감사합니다.

**Reply by Choi-JJunho** [reply]
액세스 권한이 없네요 ㅠㅠ
<img width="478" alt="image" src="https://github.com/BCSDLab/KOIN_API_V2/assets/49794401/409a0bb0-33d2-4eae-8620-bcabb12ad120">

**Reply by Choon0414** [reply]
권한 수정했습니다!

**Comment by kwoo28** [comment]
https://velog.io/@kwoo28/%EC%98%A8%EB%B3%B4%EB%94%A9-%EA%B8%B0%EB%8A%A5-%EB%B6%84%EC%84%9D%ED%95%98%EA%B8%B0
데이터베이스에 대한 연결을 datasource에서 어덯게 주입받는건지 잘 모르겠습니다.. DB에 대한 정보를 어디서 찾아서 연결하는걸까요.

**Reply by Choi-JJunho** [reply]
테스트에서 말씀하시는걸까요?
